presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java

```diff
@@ -14,86 +14,196 @@
-        StatisticRange rightRange = StatisticRange.from(rightStats);
+        superset.getSymbolsWithKnownStatistics().forEach(symbol -> {
+            SymbolStatsEstimate supersetSymbolStats = superset.getSymbolStatistics(symbol);
+            SymbolStatsEstimate leftSymbolStats = left.getSymbolStatistics(symbol);
+            SymbolStatsEstimate rightSymbolStats = right.getSymbolStatistics(symbol);
+            SymbolStatsEstimate intersectSymbolStats = intersect.getSymbolStatistics(symbol);
 
-        double nullsCountLeft = leftStats.getNullsFraction() * leftRowCount;
-        double nullsCountRight = rightStats.getNullsFraction() * rightRowCount;
-        double totalSizeLeft = (leftRowCount - nullsCountLeft) * leftStats.getAverageRowSize();
-        double totalSizeRight = (rightRowCount - nullsCountRight) * rightStats.getAverageRowSize();
-        double newNullsFraction = Math.max(nullsCountLeft - nullsCountRight, 0.0) / newRowCount;
-        double newNonNullsRowCount = newRowCount * (1.0 - newNullsFraction);
+            SymbolStatsEstimate.Builder newSymbolStats = SymbolStatsEstimate.builder();
 
-        StatisticRange range = leftRange;
-        double newDistinctValuesCount = leftStats.getDistinctValuesCount();
-        double leftValuesPerDistinctValue = leftRowCount * (1.0 - leftStats.getNullsFraction()) / leftStats.getDistinctValuesCount();
-        double rightValuesPerDistinctValue = rightRowCount * (1.0 - rightStats.getNullsFraction()) / rightStats.getDistinctValuesCount();
-        if (leftValuesPerDistinctValue <= rightValuesPerDistinctValue) {
-            // right values cover all left values for corresponding distinct values
-            range = strategy.range(leftRange, rightRange);
-            newDistinctValuesCount = leftStats.getDistinctValuesCount() - rightStats.getDistinctValuesCount();
-        }
+            // for simplicity keep the average row size the same as in the input
+            // in most cases the average row size doesn't change after applying filters
+            newSymbolStats.setAverageRowSize(supersetSymbolStats.getAverageRowSize());
 
-        return SymbolStatsEstimate.builder()
-                .setDistinctValuesCount(newDistinctValuesCount)
-                .setHighValue(range.getHigh())
-                .setLowValue(range.getLow())
-                .setAverageRowSize((totalSizeLeft - totalSizeRight) / newNonNullsRowCount)
-                .setNullsFraction(newNullsFraction)
-                .build();
+            // distinctValuesCount, range
+            // assuming distinct values are non-overlapping
+            StatisticRange range = leftSymbolStats.statisticRange().addAndSumDistinctValues(rightSymbolStats.statisticRange());
+            newSymbolStats.setStatisticsRange(range);
+            double supersetDistinctValuesCount = supersetSymbolStats.getDistinctValuesCount();
+            if (range.getDistinctValuesCount() > supersetDistinctValuesCount) {
```

Comment:
there should be check for nan in this if, see https://github.com/prestodb/presto/pull/12060/commits/a86d6dcfb0461d35a9eb4c399dfcec78e1199f0a#diff-59fa6c2264eb72e0587e7e77ce155e59R138

Reply:
From my perspective, If something under `NOT` cannot estimate NDVs, the negation code shouldn't preserve the original NDV's.